core/src/main/java/io/questdb/cairo/O3PartitionJob.java

```diff
@@ -1276,69 +1281,145 @@ private static void mergeRowGroup(
             RowGroupBuffers rowGroupBuffers,
             int rowGroupIndex,
             int timestampIndex,
-            int timestampColumnType,
+            long partitionTimestamp,
             long mergeRangeLo,
             long mergeRangeHi,
             TableRecordMetadata tableWriterMetadata,
-            long srcOooBatchRowSize
+            long srcOooBatchRowSize,
+            long dedupColSinkAddr
     ) {
+        // decode column chunks for the row group in advance
         columnsIdsAndTypes.clear();
-        columnsIdsAndTypes.add(timestampIndex);
-        columnsIdsAndTypes.add(timestampColumnType);
+        int timestampColumnChunkIndex = -1;
+        final int columnCount = tableWriterMetadata.getColumnCount();
+        for (int column = 0; column < columnCount; column++) {
+            int columnType = tableWriterMetadata.getColumnType(column);
+            if (columnType < 0) {
+                continue;
+            }
+            if (column == timestampIndex) {
+                timestampColumnChunkIndex = (int) columnsIdsAndTypes.size() / 2;
+            }
+            columnsIdsAndTypes.add(column);
+            columnsIdsAndTypes.add(columnType);
+        }
 
         final int rowGroupSize = decoder.metadata().rowGroupSize(rowGroupIndex);
         decoder.decodeRowGroup(rowGroupBuffers, columnsIdsAndTypes, rowGroupIndex, 0, rowGroupSize);
-        final long timestampDataPtr = rowGroupBuffers.getChunkDataPtr(0);
 
+        assert timestampColumnChunkIndex > -1;
+        final long timestampDataPtr = rowGroupBuffers.getChunkDataPtr(timestampColumnChunkIndex);
         long mergeBatchRowCount = mergeRangeHi - mergeRangeLo + 1;
         long mergeRowCount = mergeBatchRowCount + rowGroupSize;
+        long duplicateCount = 0;
 
-        final long timestampMergeIndexSize = mergeRowCount * TIMESTAMP_MERGE_ENTRY_BYTES;
-        final long timestampMergeIndexAddr = createMergeIndex(
-                timestampDataPtr,
-                sortedTimestampsAddr,
-                0,
-                rowGroupSize - 1,
-                mergeRangeLo,
-                mergeRangeHi,
-                timestampMergeIndexSize
-        );
+        long timestampMergeIndexSize = mergeRowCount * TIMESTAMP_MERGE_ENTRY_BYTES;
+        long timestampMergeIndexAddr;
+        if (!tableWriter.isDeduplicationEnabled()) {
+            timestampMergeIndexAddr = createMergeIndex(
+                    timestampDataPtr,
+                    sortedTimestampsAddr,
+                    0,
+                    rowGroupSize - 1,
+                    mergeRangeLo,
+                    mergeRangeHi,
+                    timestampMergeIndexSize
+            );
+        } else {
+            final DedupColumnCommitAddresses dedupCommitAddresses = tableWriter.getDedupCommitAddresses();
+            final long dedupRows;
+            timestampMergeIndexAddr = Unsafe.malloc(timestampMergeIndexSize, MemoryTag.NATIVE_O3);
```

Comment:
`try` block should start from the next line, so that we make sure to free `timestampMergeIndexAddr` if something goes wrong. Currently we only start it at L1400 after we did a lot of stuff that may throw.

Reply:
addressed